### ISSUE TITLE
map_rect with fixed integer and real data across shards

### ISSUE THREAD

**Issue by bob-carpenter** [stan-dev/math#1494]
## Description

Currently the integer and real data vary by shard in calls to `map_rect`.  In some situations, there is fixed integer or real data per shard.  Adding extra arguments for those would allow them to be shared.

This request is transcribed from [this forum post](https://discourse.mc-stan.org/t/map-rect-wish-list/12225) by Robert Kubinec, the body of which is:

> I have a proposal for what I hope would be a small change to the map_rect function, which I have come to use a lot. Currently the function takes an array of integers, an array of reals, each varying by shard, an array of shard-varying parameters and a vector of non-varying parameters.
>
> What would be really great is a vector of non-varying integers. The reason is because with more complicated models, a number of parameters need to be passed to the function to determine what kind of model to run, etc. While it is possible to append these on to the varying integer, the code would be much cleaner/logical with a separate vector.

## Example

`map_rect(...., data int[] fixed_i, data real[] fixed_r)` 

#### Current Version:
v3.0.0

**Comment by wds15** [comment]
Thanks for making the issue. In case we do this, then we should also have non-varying real data for consistency.

**Comment by bob-carpenter** [comment]
> we should also have non-varying real data for consistency.

I agree.  It wasn't in the original Discourse post, but it's in the issue title and in the signature I'm proposing.

**Comment by saudiwin** [comment]
Thanks so much guys! I didn't propose non-varying reals as they can only be primarily used as data, but I agree for consistency purposes it makes sense as I'm sure there's a use case where that would matter.

**Comment by bob-carpenter** [comment]
It's easy to create issues---we'll need to get someone to implement this.  It shouldn't be too hard but I don't think it'll land in the next release.

**Comment by wds15** [comment]
Yup. It will probably be easy to implement, but it's still plain work to plumb all the testing going with it.
